packages/compass/src/main/auto-update-manager.ts

```diff
@@ -1,12 +1,233 @@
+  Initial = 'initial',
+  Disabled = 'disabled',
+  CheckingForUpdates = 'checking-for-updates',
+  NoUpdateAvailable = 'no-update-available',
+  UpdateAvailable = 'update-available',
+  UpdateDismissed = 'update-dismissed',
+  DownloadingUpdate = 'downloading-update',
+  DownloadingError = 'downloading-error',
+  ReadyToUpdate = 'ready-to-update',
+  Restarting = 'restarting',
+  RestartDismissed = 'restart-dismissed',
+}
+
+const FOUR_HOURS = 1000 * 60 * 60 * 4;
+const THIRTY_SECONDS = 30_000;
+
+type StateUpdateAction = (
+  this: AbortSignal,
+  updateManager: typeof CompassAutoUpdateManager,
+  ...args: any[]
+) => void | Promise<void>;
+
+const checkForUpdates: StateUpdateAction = async function checkForUpdates(
+  updateManager
+) {
+  log.info(
+    mongoLogId(1001000135),
+    'AutoUpdateManager',
+    'Checking for updates ...'
+  );
+  const updateInfo = await updateManager.checkForUpdate();
+  if (this.aborted) {
+    return;
+  }
+  if (updateInfo) {
+    updateManager.setState(AutoUpdateManagerState.UpdateAvailable, updateInfo);
+  } else {
+    updateManager.setState(AutoUpdateManagerState.NoUpdateAvailable);
+  }
+};
+
+const disableAutoUpdates: StateUpdateAction = function disableAutoUpdates() {
+  log.info(
+    mongoLogId(1_001_000_148),
+    'AutoUpdateManager',
+    'Disabling auto updates'
+  );
+};
+
+/**
+ * State update map defined as `{ fromState: toState() }` where toState is the
+ * action that needs to happen on state transition
+ */
+const STATE_UPDATE: Partial<
+  Record<
+    AutoUpdateManagerState,
+    Partial<
+      Record<
+        AutoUpdateManagerState,
+        (
+          this: AbortSignal,
+          updateManager: typeof CompassAutoUpdateManager,
+          ...extraArgs: any[]
+        ) => void | Promise<void>
+      >
+    >
+  >
+> = {
+  [AutoUpdateManagerState.Initial]: {
+    [AutoUpdateManagerState.CheckingForUpdates]: checkForUpdates,
+    [AutoUpdateManagerState.Disabled]: disableAutoUpdates,
+  },
+  [AutoUpdateManagerState.Disabled]: {
+    [AutoUpdateManagerState.CheckingForUpdates]: checkForUpdates,
+  },
+  [AutoUpdateManagerState.NoUpdateAvailable]: {
+    [AutoUpdateManagerState.CheckingForUpdates]: checkForUpdates,
+    [AutoUpdateManagerState.Disabled]: disableAutoUpdates,
+  },
+  [AutoUpdateManagerState.CheckingForUpdates]: {
+    [AutoUpdateManagerState.UpdateAvailable]: async (
+      updateManager,
+      updateInfo: { from: string; to: string }
+    ) => {
+      log.info(mongoLogId(1001000127), 'AutoUpdateManager', 'Update available');
+
+      const answer = await dialog.showMessageBox({
+        icon: COMPASS_ICON,
+        title: 'New version available',
+        message: 'A new version of Compass is available to install',
+        detail: `Compass ${updateInfo.to} is available – you are currently using ${updateInfo.from}. Would you like to download and install it now?`,
+        buttons: ['Install', 'Ask me later'],
+        cancelId: 1,
+      });
+
```

Comment:
Should this also have `if (this.aborted) return;` here?

Reply:
Yeah, good catch, thank you! I'll add some more tests around this part of code too I think